### ISSUE TITLE
[General Issue] How to export Panoptic DeepLab?

### ISSUE THREAD

**Issue by ghost** [PaddlePaddle/PaddleSeg#2244]
Thanks for your issue. To help us solve the issue better, please provide following information：
 1. PaddleSeg version: PaddleSeg release/2.5
 2. PaddlePaddle version: PaddlePaddle 2.3.0 (ectually, I use docker image "registry.baidubce.com/paddlepaddle/paddle:2.3.0")
 3. Operation system: Linux
 4. Python version: Python 3.7.13
 5. CUDA/cuDNN version: None (I use cpu version of docker image)
 6. Additional context: None (Just see below).

Hi there! Thanks for reply my question!

Firstly, I try to export the model of PanopticDeepLab, which is the implementation of Panoptic Deeplab algorithm, based on PaddlePaddle refer to https://github.com/PaddlePaddle/PaddleSeg/tree/release/2.5/contrib/PanopticDeepLab

Secondly, I leverage the script, export.py( https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.5/export.py ),  to export the model by following command:

```bash
cd /path/to/PaddleSeg-release-2.5

python3 export.py   \
  --config /data/PaddleSeg-release-2.5/contrib/PanopticDeepLab/configs/panoptic_deeplab/panoptic_deeplab_resnet50_os32_cityscapes_1025x513_bs8_90k_lr00005.yml \
  --model_path /data/model/paddle_seg/pnoptic_segmentation/panoptic_deeplab_resnet50_os32_cityscapes_1025x513_bs8_90k_lr00005/model.pdparams \
  --save_dir /data/output/
```

But the script outputs these errors:

```text
Traceback (most recent call last):
  File "export.py", line 144, in <module>
    main(args)
  File "export.py", line 99, in main
    net = cfg.model
  File "/data/PaddleSeg-release-2.5/paddleseg/cvlibs/config.py", line 322, in model
    if hasattr(self.train_dataset_class, 'NUM_CLASSES'):
  File "/data/PaddleSeg-release-2.5/paddleseg/cvlibs/config.py", line 352, in train_dataset_class
    return self._load_component(dataset_type)
  File "/data/PaddleSeg-release-2.5/paddleseg/cvlibs/config.py", line 384, in _load_component
    'The specified component was not found {}.'.format(com_name))
RuntimeError: The specified component was not found CityscapesPanoptic.
``` 

Finally, please answer my question, and teach me how to solve this problem. Thank you!

best wish.

**Comment by wuyefeilin** [comment]
Sorry, there are not export.py in `contrib/PanopticDeepLab`, if you want to export the model, you copy the `export.py` to  the directory, and may be some need to update.

**Comment by ghost** [comment]
> Sorry, there are not export.py in `contrib/PanopticDeepLab`, if you want to export the model, you copy the `export.py` to the directory, and may be some need to update.

Yes, I see there is no export.py in "contrib/PanopticDeepLab/", hence I specify absolute file path in the command.

But I will try your solution.

**Comment by ghost** [comment]
> Sorry, there are not export.py in `contrib/PanopticDeepLab`, if you want to export the model, you copy the `export.py` to the directory, and may be some need to update.

I tried your solutuion (copy export.py to contrib/PanopticDeepLab) and re-run that command:

```bash
cd /path/to/PaddleSeg-release-2.5

cp export.py contrib/PanopticDeepLab/

python3 export.py \
  --config configs/panoptic_deeplab/panoptic_deeplab_resnet50_os32_cityscapes_1025x513_bs8_90k_lr00005.yml \
  --model_path /data/model/paddle_seg/pnoptic_segmentation/panoptic_deeplab_resnet50_os32_cityscapes_1025x513_bs8_90k_lr00005/model.pdparams \
  --save_dir /data/output/
```

Still, there are RuntimeError:

```text
grep: warning: GREP_OPTIONS is deprecated; please use an alias or script
/usr/local/lib/python3.7/dist-packages/paddleseg-2.5.0-py3.7.egg/paddleseg/models/losses/rmi_loss.py:78: DeprecationWarning: invalid escape sequence \i
/usr/local/lib/python3.7/dist-packages/paddleseg-2.5.0-py3.7.egg/paddleseg/models/losses/rmi_loss.py:78: DeprecationWarning: invalid escape sequence \i
Traceback (most recent call last):
  File "export.py", line 144, in <module>
    main(args)
  File "export.py", line 99, in main
    net = cfg.model
  File "/usr/local/lib/python3.7/dist-packages/paddleseg-2.5.0-py3.7.egg/paddleseg/cvlibs/config.py", line 322, in model
  File "/usr/local/lib/python3.7/dist-packages/paddleseg-2.5.0-py3.7.egg/paddleseg/cvlibs/config.py", line 352, in train_dataset_class
  File "/usr/local/lib/python3.7/dist-packages/paddleseg-2.5.0-py3.7.egg/paddleseg/cvlibs/config.py", line 384, in _load_component
RuntimeError: The specified component was not found CityscapesPanoptic.
```

I survey some issues of this repo. I found that this problem seems owing to the dataset missing? But that issuse (Sorry I can not find it again) said that we can export the model without a dataset after release/v2.0 .

At this time, what is the main reason for my problem?

**Comment by ghost** [comment]
Carry on the last reply.

I think my problem is the dataset missing, which is [CityScapes](https://www.cityscapes-dataset.com/). 

So I downloaded and [prepared](https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/preparation/createPanopticImgs.py) the dataset like below:

```
cityscapes/
|--gtFine/
|  |--train/
|  |  |--aachen/
|  |  |  |--*_color.png, *_instanceIds.png, *_labelIds.png, *_polygons.json,
|  |  |  |--*_labelTrainIds.png
|  |  |  |--...
|  |--val/
|  |--test/
|  |--cityscapes_panoptic_train_trainId.json
|  |--cityscapes_panoptic_train_trainId/
|  |  |-- *_panoptic.png
|  |--cityscapes_panoptic_val_trainId.json
|  |--cityscapes_panoptic_val_trainId/
|  |  |--  *_panoptic.png
|--leftImg8bit/
|  |--train/
|  |--val/
|  |--test/
```

Unfortunately, when I run the export command:

```bash
cd PaddleSeg-release-2.5/contrib/PanopticDeepLab

python3 export.py \
  --config configs/panoptic_deeplab/panoptic_deeplab_resnet50_os32_cityscapes_1025x513_bs8_90k_lr00005.yml \
  --model_path /data/model/paddle_seg/pnoptic_segmentation/panoptic_deeplab_resnet50_os32_cityscapes_1025x513_bs8_90k_lr00005/model.pdparams \
  --save_dir ./output/result
```

An error occurred:

```
grep: warning: GREP_OPTIONS is deprecated; please use an alias or script
/usr/local/lib/python3.7/dist-packages/paddleseg-2.5.0-py3.7.egg/paddleseg/models/losses/rmi_loss.py:78: DeprecationWarning: invalid escape sequence \i
/usr/local/lib/python3.7/dist-packages/paddleseg-2.5.0-py3.7.egg/paddleseg/models/losses/rmi_loss.py:78: DeprecationWarning: invalid escape sequence \i
Traceback (most recent call last):
  File "export.py", line 144, in <module>
    main(args)
  File "export.py", line 99, in main
    net = cfg.model
  File "/usr/local/lib/python3.7/dist-packages/paddleseg-2.5.0-py3.7.egg/paddleseg/cvlibs/config.py", line 322, in model
  File "/usr/local/lib/python3.7/dist-packages/paddleseg-2.5.0-py3.7.egg/paddleseg/cvlibs/config.py", line 352, in train_dataset_class
  File "/usr/local/lib/python3.7/dist-packages/paddleseg-2.5.0-py3.7.egg/paddleseg/cvlibs/config.py", line 384, in _load_component
RuntimeError: The specified component was not found CityscapesPanoptic.
```

Yes, the error is as same as our beginning. 

In other words, this errors (RuntimeError: The specified component was not found CityscapesPanoptic.) is does not matter with the dataset missing.

Please help to solve this problem~

Best wishes!

**Comment by wuyefeilin** [comment]
You should  import the CityscapesPanoptic class from datasets in export.py.

What project are you using about this model?  Or may be you just learning about it?

**Comment by ghost** [comment]
> You should import the CityscapesPanoptic class from datasets in export.py.
> 
> What project are you using about this model? Or maybe you just learning about it?

Thanks for your solution, I will try it!

I am not learning "exporting Panoptic DeepLab to ONNX".

This issue is involved in an enterprise project. Our company has a real-world scene, which is going to deploy the Panoptic Segmentation models on Edge devices.

Now, the project is in the early research stage. 

We would like PaddlePaddle to be applied to our technology stacks.

**Comment by ghost** [comment]
According to suggestions of the developers of PaddleSeg. This problem can be solved by following steps:

Step 1, copy export.py to `contrib/PanopticDeepLab`;

Step 2, add the following codes to `contrib/PanopticDeepLab/export.py`:

```python
from datasets import CityscapesPanoptic
from models import PanopticDeepLab
```

Step 3, download pre-trained model file:

```bash
mkdir -p /data/model/paddle_seg/pnoptic_segmentation/panoptic_deeplab_resnet50_os32_cityscapes_1025x513_bs8_90k_lr00005/

cd /data/model/paddle_seg/pnoptic_segmentation/panoptic_deeplab_resnet50_os32_cityscapes_1025x513_bs8_90k_lr00005/

wget https://bj.bcebos.com/paddleseg/dygraph/pnoptic_segmentation/panoptic_deeplab_resnet50_os32_cityscapes_1025x513_bs8_90k_lr00005/model.pdparams
```

Step 4, run `contrib/PanopticDeepLab/export.py` by the following command:

```bash
cd /path/to/contrib/PanopticDeepLab

python3 export.py   \
    --config configs/panoptic_deeplab/panoptic_deeplab_resnet50_os32_cityscapes_1025x513_bs8_90k_lr00005.yml \
    --model_path /data/model/paddle_seg/pnoptic_segmentation/panoptic_deeplab_resnet50_os32_cityscapes_1025x513_bs8_90k_lr00005/model.pdparams \
    --save_dir ./output/model \
    --input_shape 1 3 1024 1024
```

Finally, exported model files can be found at `contrib/PanopticDeepLab/output/model`:

```text
deploy.yaml
model.pdiparams
model.pdiparams.info
model.pdmodel
```
